### PR TITLE
Make page & candidate selections as round-robin

### PR DESCRIPTION
**Pull request by sun409**

### CODE REVIEW

**Comment by tumashu** [comment]
这个pr包含的无关内容太多了吧？

**Comment by sun409** [comment]
I removed meaningless translation of Chinese comments. It should be clearer now.

File: pyim.el [hunk 1]
```diff
@@ -1310,7 +1313,7 @@ dcache 文件的方法让 pyim 正常工作。")
     (while (< i 256)
       (define-key map (vector i) 'pyim-self-insert-command)
       (setq i (1+ i)))
-    (dolist (i (number-sequence ?1 ?9))
+    (dolist (i (number-sequence ?0 ?9))
```
Comment (sun409) [inline pyim.el]: Make number 0 and 1..9 can be used to select word

File: pyim.el [hunk 2]
```diff
@@ -1479,14 +1482,12 @@ dcache 文件的方法让 pyim 正常工作。")
...
-          output
-        (delq ""
-              (mapcar (lambda (x)
-                        (replace-regexp-in-string "|" "" x))
-                      (or output '())))))))
+                   (gethash pinyin pyim-pinyin2cchar-cache1))
+               (gethash pinyin pyim-pinyin2cchar-cache3))))
+         (delete "" output)
+         (if include-seperator
+             output
```
Comment (sun409) [inline pyim.el]: Remove empty string or there will be an empty candidate in the place of "|".

File: pyim.el [hunk 3]
```diff
@@ -2768,7 +2770,7 @@ pyim 会使用 emacs overlay 机制在 *待输入buffer* 光标处高亮显示
 
 细节信息请参考 `pyim-page-refresh' 的 docstring."
   (let ((pos (min (length pyim-candidates) pyim-candidate-position)))
-    (1+ (- pos (mod pos pyim-page-length)))))
+       (1+ (* (/ (1- pos) pyim-page-length) pyim-page-length))))
```
Comment (sun409) [inline pyim.el]: The original one will be wrong when pos is the multiple of pyim-page-length.

File: pyim.el [hunk 4]
```diff
@@ -2905,8 +2907,12 @@ minibuffer 原来显示的信息和 pyim 选词框整合在一起显示
...
         (pyim-terminate-translation))
     (let ((new (+ pyim-candidate-position (* pyim-page-length arg) 1)))
-      (setq pyim-candidate-position (if (> new 0) new 1)
-            pyim-candidate-position (pyim-page-start))
+      (setq maxpos (length pyim-candidates))
+      (setq pyim-candidate-position
+            (if (> new 0)
+                (if (> new maxpos) 1 new)
+                maxpos)
+        pyim-candidate-position (pyim-page-start))
```
Comment (sun409) [inline pyim.el]: Make it wrap around.

File: pyim.el [hunk 5]
```diff
@@ -2921,7 +2927,11 @@ minibuffer 原来显示的信息和 pyim 选词框整合在一起显示
         (pyim-outcome-handle 'last-char)
         (pyim-terminate-translation))
     (let ((new (+ pyim-candidate-position arg)))
-      (setq pyim-candidate-position (if (> new 0) new 1))
+      (setq len (length pyim-candidates))
+      (setq pyim-candidate-position
+        (if (>= len new)
+            (if (> new 0) new len)
+            1))
```
Comment (sun409) [inline pyim.el]: Make it wrap around.

File: pyim.el [hunk 6]
```diff
@@ -3338,9 +3348,10 @@ minibuffer 原来显示的信息和 pyim 选词框整合在一起显示
             (pyim-outcome-handle 'last-char)
             (pyim-terminate-translation))
         (let ((index (if (numberp n)
-                         (- n 1)
+                          (- n 1)
                        (- last-command-event ?1)))
               (end (pyim-page-end)))
+          (if (= index -1) (setq index 9) nil)
```
Comment (sun409) [inline pyim.el]: Special case for ?0 since its meaning should be 10.

**Comment by tumashu** [comment]
我建议单独开一个branch，重新做一个干净的pr，只包含需要合并的东西

**Pull request closed without merging**